numpy/core/tests/test_numerictypes.py

```diff
@@ -499,3 +499,25 @@ def test_platform_dependent_aliases(self):
             assert_('int64' in np.int_.__doc__)
         elif np.int64 is np.longlong:
             assert_('int64' in np.longlong.__doc__)
+
+
+def test_scalar_type_names():
+    # gh-9799
+    numeric_types = [
+        np.byte, np.short, np.intc, np.int_, np.longlong,
+        np.ubyte, np.ushort, np.uintc, np.uint, np.ulonglong,
+        np.half, np.single, np.double, np.longdouble,
+        np.csingle, np.cdouble, np.clongdouble,
+    ]
+
+    # none of the above may be aliases for each other
+    int_ids = [id(t) for t in numeric_types]
+    assert_equal(len(set(int_ids)), len(int_ids))
+
+    # names must be unique
+    int_names = [t.__name__ for t in numeric_types]
+    assert_equal(len(set(int_names)), len(int_names))
+
+    # names should actually correspond to where the type is
+    for t in numeric_types:
+        assert getattr(np, t.__name__) is t
```

Comment:
OK, this may be silly, but can we add to the test:
```
for t in numeric_types:
    assert np.dtype(t.__name__).type is t
```
Since I think this is true with your changes (also then I do not have to check the above list manually for correctness).

Reply:
That sounds like a good idea to me